base/array.jl

```diff
@@ -2672,3 +2672,33 @@ function intersect(v::AbstractVector, r::AbstractRange)
     return vectorfilter(T, _shrink_filter!(seen), common)
 end
 intersect(r::AbstractRange, v::AbstractVector) = intersect(v, r)
+
+
+_collect_n(itr, ::Val{0}) = error()
+_collect_n(itr, ::Val{0}, st) = ((), st)
+function _collect_n(itr, ::Val{N}, st...) where {N}
+    tmp = iterate(itr, st...)
+    if tmp === nothing
+        error("Iterator does not contain enough elements for the given variables.")
+    end
+    first, st′ = tmp
+    tail, st′′ = _collect_n(itr, Val(N-1), st′)
+    return (first, tail...), st′′
+end
+
+function split_rest(itr, ::Val{N}, st...) where {N}
+    if IteratorSize(itr) == IsInfinite()
+        error("Can't split an infinite iterator in the middle.")
+    end
+    last_n, st′ = _collect_n(itr, Val(N), st...)
+    front = Vector{@default_eltype(itr)}()
+    while true
+        tmp = iterate(itr, st′)
+        tmp === nothing && break
+        xᵢ, st′ = tmp
+        push!(front, first(last_n))
+        last_n = (tail(last_n)..., xᵢ)
+    end
```

Comment:
Is there any case where this works better than just collecting the rest of the iterator and splitting it into two vectors?

Reply:
It seems like the static version using tuples is ever so slightly faster for very small arrays, but a lot slower for larger sizes:
```
julia> results = run(suite)
2-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "split_rest_dynamic" => 9-element BenchmarkTools.BenchmarkGroup:
	  tags: []
	  "(len = 1000, N = 16)" => Trial(676.300 ns)
	  "(len = 100, N = 2)" => Trial(98.819 ns)
	  "(len = 1000, N = 2)" => Trial(662.985 ns)
	  "(len = 1000, N = 8)" => Trial(666.324 ns)
	  "(len = 10, N = 2)" => Trial(69.341 ns)
	  "(len = 10, N = 8)" => Trial(67.838 ns)
	  "(len = 100, N = 16)" => Trial(100.039 ns)
	  "(len = 100, N = 8)" => Trial(98.501 ns)
	  "(len = 5, N = 2)" => Trial(67.053 ns)
  "split_rest_static" => 9-element BenchmarkTools.BenchmarkGroup:
	  tags: []
	  "(len = 1000, N = 16)" => Trial(8.870 μs)
	  "(len = 100, N = 2)" => Trial(638.733 ns)
	  "(len = 1000, N = 2)" => Trial(6.118 μs)
	  "(len = 1000, N = 8)" => Trial(7.281 μs)
	  "(len = 10, N = 2)" => Trial(83.375 ns)
	  "(len = 10, N = 8)" => Trial(63.285 ns)
	  "(len = 100, N = 16)" => Trial(819.295 ns)
	  "(len = 100, N = 8)" => Trial(702.299 ns)
	  "(len = 5, N = 2)" => Trial(60.169 ns)
```
So just collecting the whole iterator first probably makes the most sense here.